### ISSUE TITLE
Pyright doesn't detect wrong argument when the annotation involves a type variable

### ISSUE THREAD

**Issue by danpascu** [microsoft/pyright#8192]
With this code:

```py

from typing import ClassVar

class Base: ...

class Root[T: Base]:
    _type: ClassVar[type[Base] | None] = None
    _name: ClassVar[str | None] = None

    def __init_subclass__(cls, element_type: type[T] | None = None, name: str | None = None, **kw: object) -> None:
        super().__init_subclass__(**kw)
        cls._type = element_type
        cls._name = name

# There is no error here about element_type being of the wrong type (should be a subtype of Base | None)
class Test1(Root, element_type=int):
    pass

# This is correctly reported as name cannot be Literal[1], it must be str | None
class Test2(Root, name=1):
    pass
```

I get no error for Test1 where element_type should be a subtype of Base | None, while I correctly get an error for Test2 where name should be str | None.

It seems that if a type variable is involved in the annotation, the error is not detected anymore.

This is with command line pyright 1.1.368

**Comment by erictraut** [comment]
Pyright's behavior here is correct. `Test1` subclasses from `Root` which is a generic class that requires a single type argument. Since you haven't provided any such type argument, it defaults to `Any`. That means you're subclassing from `Root[Any]`, and `T` has the type `Any`.

**Comment by danpascu** [comment]
But how can that default to `Any` when `T` is bound by `Base`? Shouldn't that be at least `Root[Base]`?

I though explicitly specifying the type is not needed when you provide at least one of the arguments that are annotated with the generic type and it can infer it from that.

I know I used this in the past like this and it was correctly detected:

```py
class Base: ...

class Root[T: Base]:
    _type: type[T] | None = None

    def __init__(self, _type: type[T], /) -> None:
        self._type = _type


r = Root(int)  # error: type "int" is incompatible with type "Base"
```

Here I don't need to call it like `Root[int](int)`, `Root(int)` is enough to infer the type of `T` and give me an error because it doesn't conform to the bound.

I agree the context is a bit different (this one instantiates the class, while the other is sub-classing it), but they both end up calling a method with an argument annotated with the generic type.

I can provide the type when sub-classing, but I'd like to understand why these 2 are handled differently and in one case type inference happens and it gives an error, while in the other case it just assumes `Any` and moves on.

**Comment by erictraut** [comment]
> Shouldn't that be at least Root[Base]?

No, that's not how it works in the Python type system. For details, refer to the [generics chapter of the Python typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html). Any unspecified type arguments default to `Any` unless the type parameter has been defined with an explicit default value. This is a new capability introduced in Python 3.13. For details, refer to [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html#defaults-for-type-parameters).

> I though explicitly specifying the type is not needed when you provide at least one of the arguments that are annotated with the generic type and it can infer it from that.

Type inference of class-scoped type variables is performed only for constructor calls. For more information refer to the [constructor chapter of the typing spec](https://typing.readthedocs.io/en/latest/spec/constructors.html).
